// @cryptoalgebra/integral-core/dist/src/core/typechain/factories/contracts/interfaces/vault/IAlgebraVaultFactory__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IAlgebraVaultFactory__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "address",
                name: "pool",
                type: "address",
            },
            {
                internalType: "address",
                name: "creator",
                type: "address",
            },
            {
                internalType: "address",
                name: "deployer",
                type: "address",
            },
            {
                internalType: "address",
                name: "token0",
                type: "address",
            },
            {
                internalType: "address",
                name: "token1",
                type: "address",
            },
        ],
        name: "createVaultForPool",
        outputs: [
            {
                internalType: "address",
                name: "communityFeeVault",
                type: "address",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "pool",
                type: "address",
            },
        ],
        name: "getVaultForPool",
        outputs: [
            {
                internalType: "address",
                name: "communityFeeVault",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
class IAlgebraVaultFactory__factory {
    static abi = _abi;
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.IAlgebraVaultFactory__factory = IAlgebraVaultFactory__factory;
