// @cryptoalgebra/integral-core/dist/src/core/hardhat.config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
require("@nomicfoundation/hardhat-toolbox");
require("hardhat-contract-sizer");
require("hardhat-output-validator");
require("solidity-docgen");
const hardhat_base_config_1 = __importDefault(require("../../hardhat.base.config"));
const HIGHEST_OPTIMIZER_COMPILER_SETTINGS = {
    version: '0.8.20',
    settings: {
        evmVersion: 'paris',
        optimizer: {
            enabled: true,
            runs: 1000000,
        },
        metadata: {
            bytecodeHash: 'none',
        },
    },
};
const DEFAULT_COMPILER_SETTINGS = {
    version: '0.8.20',
    settings: {
        evmVersion: 'paris',
        optimizer: {
            enabled: true,
            runs: 800,
        },
        metadata: {
            bytecodeHash: 'none',
        },
    },
};
const HIGH_COMPILER_SETTINGS = {
    version: '0.8.20',
    settings: {
        evmVersion: 'paris',
        optimizer: {
            enabled: true,
            runs: 0,
        },
        metadata: {
            bytecodeHash: 'none',
        },
    },
};
if (process.env.RUN_COVERAGE == '1') {
    /**
     * Updates the default compiler settings when running coverage.
     *
     * See https://github.com/sc-forks/solidity-coverage/issues/417#issuecomment-730526466
     */
    console.info('Using coverage compiler settings');
    const details = {
        yul: true,
        yulDetails: {
            stackAllocation: true,
        },
    };
    HIGHEST_OPTIMIZER_COMPILER_SETTINGS.settings.details = details;
    DEFAULT_COMPILER_SETTINGS.settings.details = details;
}
exports.default = {
    networks: hardhat_base_config_1.default.networks,
    etherscan: hardhat_base_config_1.default.etherscan,
    typechain: {
        outDir: 'typechain',
    },
    solidity: {
        compilers: [DEFAULT_COMPILER_SETTINGS],
        overrides: {
            'contracts/AlgebraFactory.sol': HIGHEST_OPTIMIZER_COMPILER_SETTINGS,
            'contracts/AlgebraPool.sol': HIGH_COMPILER_SETTINGS,
            'contracts/AlgebraPoolDeployer.sol': HIGH_COMPILER_SETTINGS,
        },
    },
    docgen: {
        outputDir: '../../docs/Contracts/Core',
        pages: (x, buildInfo) => {
            return `${buildInfo.relativePath}`.replace('.sol', '.md');
        },
        templates: '../../docs/doc_templates/public',
        collapseNewlines: true,
    },
    outputValidator: {
        runOnCompile: false,
        exclude: ['contracts/test'],
    },
};
