// @cryptoalgebra/integral-core/typechain/contracts/test/echidna/LiquidityMathEchidnaTest.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumberish,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedListener,
  TypedContractMethod,
} from "../../../common";

export interface LiquidityMathEchidnaTestInterface extends Interface {
  getFunction(
    nameOrSignature: "addDelta" | "checkGetAmountsForLiquidity"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "addDelta",
    values: [BigNumberish, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "checkGetAmountsForLiquidity",
    values: [BigNumberish, BigNumberish, BigNumberish, BigNumberish]
  ): string;

  decodeFunctionResult(functionFragment: "addDelta", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "checkGetAmountsForLiquidity",
    data: BytesLike
  ): Result;
}

export interface LiquidityMathEchidnaTest extends BaseContract {
  connect(runner?: ContractRunner | null): LiquidityMathEchidnaTest;
  waitForDeployment(): Promise<this>;

  interface: LiquidityMathEchidnaTestInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  addDelta: TypedContractMethod<
    [x: BigNumberish, y: BigNumberish],
    [void],
    "view"
  >;

  checkGetAmountsForLiquidity: TypedContractMethod<
    [
      bottomTick: BigNumberish,
      topTick: BigNumberish,
      liquidityDelta: BigNumberish,
      currentPrice: BigNumberish
    ],
    [void],
    "view"
  >;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "addDelta"
  ): TypedContractMethod<[x: BigNumberish, y: BigNumberish], [void], "view">;
  getFunction(
    nameOrSignature: "checkGetAmountsForLiquidity"
  ): TypedContractMethod<
    [
      bottomTick: BigNumberish,
      topTick: BigNumberish,
      liquidityDelta: BigNumberish,
      currentPrice: BigNumberish
    ],
    [void],
    "view"
  >;

  filters: {};
}
