// @cryptoalgebra/integral-core/typechain/contracts/test/echidna/TickTreeSearchEchidnaTest.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumberish,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedListener,
  TypedContractMethod,
} from "../../../common";

export interface TickTreeSearchEchidnaTestInterface extends Interface {
  getFunction(
    nameOrSignature: "checkNextInitializedTickInvariants" | "toggleTick"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "checkNextInitializedTickInvariants",
    values: [BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "toggleTick",
    values: [BigNumberish]
  ): string;

  decodeFunctionResult(
    functionFragment: "checkNextInitializedTickInvariants",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "toggleTick", data: BytesLike): Result;
}

export interface TickTreeSearchEchidnaTest extends BaseContract {
  connect(runner?: ContractRunner | null): TickTreeSearchEchidnaTest;
  waitForDeployment(): Promise<this>;

  interface: TickTreeSearchEchidnaTestInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  checkNextInitializedTickInvariants: TypedContractMethod<
    [tick: BigNumberish],
    [void],
    "view"
  >;

  toggleTick: TypedContractMethod<[tick: BigNumberish], [void], "nonpayable">;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "checkNextInitializedTickInvariants"
  ): TypedContractMethod<[tick: BigNumberish], [void], "view">;
  getFunction(
    nameOrSignature: "toggleTick"
  ): TypedContractMethod<[tick: BigNumberish], [void], "nonpayable">;

  filters: {};
}
