// @cryptoalgebra/integral-core/typechain/contracts/test/echidna/TickOverflowSafetyEchidnaTest.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumberish,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedListener,
  TypedContractMethod,
} from "../../../common";

export interface TickOverflowSafetyEchidnaTestInterface extends Interface {
  getFunction(
    nameOrSignature:
      | "increaseTotalFeeGrowth0Token"
      | "increaseTotalFeeGrowth1Token"
      | "moveToTick"
      | "setPosition"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "increaseTotalFeeGrowth0Token",
    values: [BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "increaseTotalFeeGrowth1Token",
    values: [BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "moveToTick",
    values: [BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "setPosition",
    values: [BigNumberish, BigNumberish, BigNumberish]
  ): string;

  decodeFunctionResult(
    functionFragment: "increaseTotalFeeGrowth0Token",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "increaseTotalFeeGrowth1Token",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "moveToTick", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "setPosition",
    data: BytesLike
  ): Result;
}

export interface TickOverflowSafetyEchidnaTest extends BaseContract {
  connect(runner?: ContractRunner | null): TickOverflowSafetyEchidnaTest;
  waitForDeployment(): Promise<this>;

  interface: TickOverflowSafetyEchidnaTestInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  increaseTotalFeeGrowth0Token: TypedContractMethod<
    [amount: BigNumberish],
    [void],
    "nonpayable"
  >;

  increaseTotalFeeGrowth1Token: TypedContractMethod<
    [amount: BigNumberish],
    [void],
    "nonpayable"
  >;

  moveToTick: TypedContractMethod<[target: BigNumberish], [void], "nonpayable">;

  setPosition: TypedContractMethod<
    [
      bottomTick: BigNumberish,
      topTick: BigNumberish,
      liquidityDelta: BigNumberish
    ],
    [void],
    "nonpayable"
  >;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "increaseTotalFeeGrowth0Token"
  ): TypedContractMethod<[amount: BigNumberish], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "increaseTotalFeeGrowth1Token"
  ): TypedContractMethod<[amount: BigNumberish], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "moveToTick"
  ): TypedContractMethod<[target: BigNumberish], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "setPosition"
  ): TypedContractMethod<
    [
      bottomTick: BigNumberish,
      topTick: BigNumberish,
      liquidityDelta: BigNumberish
    ],
    [void],
    "nonpayable"
  >;

  filters: {};
}
