// @cryptoalgebra/integral-core/typechain/factories/contracts/interfaces/plugin/IAlgebraPluginFactory__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Interface, type ContractRunner } from "ethers";
import type {
  IAlgebraPluginFactory,
  IAlgebraPluginFactoryInterface,
} from "../../../../contracts/interfaces/plugin/IAlgebraPluginFactory";

const _abi = [
  {
    inputs: [
      {
        internalType: "address",
        name: "plugin",
        type: "address",
      },
      {
        internalType: "address",
        name: "pool",
        type: "address",
      },
      {
        internalType: "address",
        name: "deployer",
        type: "address",
      },
    ],
    name: "afterCreatePoolHook",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "pool",
        type: "address",
      },
      {
        internalType: "address",
        name: "creator",
        type: "address",
      },
      {
        internalType: "address",
        name: "deployer",
        type: "address",
      },
      {
        internalType: "address",
        name: "token0",
        type: "address",
      },
      {
        internalType: "address",
        name: "token1",
        type: "address",
      },
      {
        internalType: "bytes",
        name: "data",
        type: "bytes",
      },
    ],
    name: "beforeCreatePoolHook",
    outputs: [
      {
        internalType: "address",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "nonpayable",
    type: "function",
  },
] as const;

export class IAlgebraPluginFactory__factory {
  static readonly abi = _abi;
  static createInterface(): IAlgebraPluginFactoryInterface {
    return new Interface(_abi) as IAlgebraPluginFactoryInterface;
  }
  static connect(
    address: string,
    runner?: ContractRunner | null
  ): IAlgebraPluginFactory {
    return new Contract(
      address,
      _abi,
      runner
    ) as unknown as IAlgebraPluginFactory;
  }
}
