// @cryptoalgebra/integral-core/typechain/factories/contracts/interfaces/callback/IAlgebraFlashCallback__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Interface, type ContractRunner } from "ethers";
import type {
  IAlgebraFlashCallback,
  IAlgebraFlashCallbackInterface,
} from "../../../../contracts/interfaces/callback/IAlgebraFlashCallback";

const _abi = [
  {
    inputs: [
      {
        internalType: "uint256",
        name: "fee0",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "fee1",
        type: "uint256",
      },
      {
        internalType: "bytes",
        name: "data",
        type: "bytes",
      },
    ],
    name: "algebraFlashCallback",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
] as const;

export class IAlgebraFlashCallback__factory {
  static readonly abi = _abi;
  static createInterface(): IAlgebraFlashCallbackInterface {
    return new Interface(_abi) as IAlgebraFlashCallbackInterface;
  }
  static connect(
    address: string,
    runner?: ContractRunner | null
  ): IAlgebraFlashCallback {
    return new Contract(
      address,
      _abi,
      runner
    ) as unknown as IAlgebraFlashCallback;
  }
}
