// @cryptoalgebra/integral-core/typechain/factories/contracts/interfaces/callback/IAlgebraMintCallback__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Interface, type ContractRunner } from "ethers";
import type {
  IAlgebraMintCallback,
  IAlgebraMintCallbackInterface,
} from "../../../../contracts/interfaces/callback/IAlgebraMintCallback";

const _abi = [
  {
    inputs: [
      {
        internalType: "uint256",
        name: "amount0Owed",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "amount1Owed",
        type: "uint256",
      },
      {
        internalType: "bytes",
        name: "data",
        type: "bytes",
      },
    ],
    name: "algebraMintCallback",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
] as const;

export class IAlgebraMintCallback__factory {
  static readonly abi = _abi;
  static createInterface(): IAlgebraMintCallbackInterface {
    return new Interface(_abi) as IAlgebraMintCallbackInterface;
  }
  static connect(
    address: string,
    runner?: ContractRunner | null
  ): IAlgebraMintCallback {
    return new Contract(
      address,
      _abi,
      runner
    ) as unknown as IAlgebraMintCallback;
  }
}
