// @cryptoalgebra/integral-core/typechain/contracts/AlgebraFactory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumberish,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  EventFragment,
  AddressLike,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedLogDescription,
  TypedListener,
  TypedContractMethod,
} from "../common";

export interface AlgebraFactoryInterface extends Interface {
  getFunction(
    nameOrSignature:
      | "CUSTOM_POOL_DEPLOYER"
      | "DEFAULT_ADMIN_ROLE"
      | "POOLS_ADMINISTRATOR_ROLE"
      | "POOL_INIT_CODE_HASH"
      | "acceptOwnership"
      | "computeCustomPoolAddress"
      | "computePoolAddress"
      | "createCustomPool"
      | "createPool"
      | "customPoolByPair"
      | "defaultCommunityFee"
      | "defaultConfigurationForPool"
      | "defaultFee"
      | "defaultPluginFactory"
      | "defaultTickspacing"
      | "getRoleAdmin"
      | "getRoleMember"
      | "getRoleMemberCount"
      | "grantRole"
      | "hasRole"
      | "hasRoleOrOwner"
      | "owner"
      | "pendingOwner"
      | "poolByPair"
      | "poolDeployer"
      | "renounceOwnership"
      | "renounceOwnershipStartTimestamp"
      | "renounceRole"
      | "revokeRole"
      | "setDefaultCommunityFee"
      | "setDefaultFee"
      | "setDefaultPluginFactory"
      | "setDefaultTickspacing"
      | "setVaultFactory"
      | "startRenounceOwnership"
      | "stopRenounceOwnership"
      | "supportsInterface"
      | "transferOwnership"
      | "vaultFactory"
  ): FunctionFragment;

  getEvent(
    nameOrSignatureOrTopic:
      | "CustomPool"
      | "DefaultCommunityFee"
      | "DefaultFee"
      | "DefaultPluginFactory"
      | "DefaultTickspacing"
      | "OwnershipTransferStarted"
      | "OwnershipTransferred"
      | "Pool"
      | "RenounceOwnershipFinish"
      | "RenounceOwnershipStart"
      | "RenounceOwnershipStop"
      | "RoleAdminChanged"
      | "RoleGranted"
      | "RoleRevoked"
      | "VaultFactory"
  ): EventFragment;

  encodeFunctionData(
    functionFragment: "CUSTOM_POOL_DEPLOYER",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "DEFAULT_ADMIN_ROLE",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "POOLS_ADMINISTRATOR_ROLE",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "POOL_INIT_CODE_HASH",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "acceptOwnership",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "computeCustomPoolAddress",
    values: [AddressLike, AddressLike, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "computePoolAddress",
    values: [AddressLike, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "createCustomPool",
    values: [AddressLike, AddressLike, AddressLike, AddressLike, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "createPool",
    values: [AddressLike, AddressLike, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "customPoolByPair",
    values: [AddressLike, AddressLike, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "defaultCommunityFee",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "defaultConfigurationForPool",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "defaultFee",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "defaultPluginFactory",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "defaultTickspacing",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "getRoleAdmin",
    values: [BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "getRoleMember",
    values: [BytesLike, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "getRoleMemberCount",
    values: [BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "grantRole",
    values: [BytesLike, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "hasRole",
    values: [BytesLike, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "hasRoleOrOwner",
    values: [BytesLike, AddressLike]
  ): string;
  encodeFunctionData(functionFragment: "owner", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "pendingOwner",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "poolByPair",
    values: [AddressLike, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "poolDeployer",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "renounceOwnership",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "renounceOwnershipStartTimestamp",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "renounceRole",
    values: [BytesLike, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "revokeRole",
    values: [BytesLike, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "setDefaultCommunityFee",
    values: [BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "setDefaultFee",
    values: [BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "setDefaultPluginFactory",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "setDefaultTickspacing",
    values: [BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "setVaultFactory",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "startRenounceOwnership",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "stopRenounceOwnership",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "supportsInterface",
    values: [BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "transferOwnership",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "vaultFactory",
    values?: undefined
  ): string;

  decodeFunctionResult(
    functionFragment: "CUSTOM_POOL_DEPLOYER",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "DEFAULT_ADMIN_ROLE",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "POOLS_ADMINISTRATOR_ROLE",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "POOL_INIT_CODE_HASH",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "acceptOwnership",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "computeCustomPoolAddress",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "computePoolAddress",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "createCustomPool",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "createPool", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "customPoolByPair",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "defaultCommunityFee",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "defaultConfigurationForPool",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "defaultFee", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "defaultPluginFactory",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "defaultTickspacing",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getRoleAdmin",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getRoleMember",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getRoleMemberCount",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "grantRole", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "hasRole", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "hasRoleOrOwner",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "owner", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "pendingOwner",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "poolByPair", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "poolDeployer",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "renounceOwnership",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "renounceOwnershipStartTimestamp",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "renounceRole",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "revokeRole", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "setDefaultCommunityFee",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setDefaultFee",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setDefaultPluginFactory",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setDefaultTickspacing",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setVaultFactory",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "startRenounceOwnership",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "stopRenounceOwnership",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "supportsInterface",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "transferOwnership",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "vaultFactory",
    data: BytesLike
  ): Result;
}

export namespace CustomPoolEvent {
  export type InputTuple = [
    deployer: AddressLike,
    token0: AddressLike,
    token1: AddressLike,
    pool: AddressLike
  ];
  export type OutputTuple = [
    deployer: string,
    token0: string,
    token1: string,
    pool: string
  ];
  export interface OutputObject {
    deployer: string;
    token0: string;
    token1: string;
    pool: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace DefaultCommunityFeeEvent {
  export type InputTuple = [newDefaultCommunityFee: BigNumberish];
  export type OutputTuple = [newDefaultCommunityFee: bigint];
  export interface OutputObject {
    newDefaultCommunityFee: bigint;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace DefaultFeeEvent {
  export type InputTuple = [newDefaultFee: BigNumberish];
  export type OutputTuple = [newDefaultFee: bigint];
  export interface OutputObject {
    newDefaultFee: bigint;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace DefaultPluginFactoryEvent {
  export type InputTuple = [defaultPluginFactoryAddress: AddressLike];
  export type OutputTuple = [defaultPluginFactoryAddress: string];
  export interface OutputObject {
    defaultPluginFactoryAddress: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace DefaultTickspacingEvent {
  export type InputTuple = [newDefaultTickspacing: BigNumberish];
  export type OutputTuple = [newDefaultTickspacing: bigint];
  export interface OutputObject {
    newDefaultTickspacing: bigint;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace OwnershipTransferStartedEvent {
  export type InputTuple = [previousOwner: AddressLike, newOwner: AddressLike];
  export type OutputTuple = [previousOwner: string, newOwner: string];
  export interface OutputObject {
    previousOwner: string;
    newOwner: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace OwnershipTransferredEvent {
  export type InputTuple = [previousOwner: AddressLike, newOwner: AddressLike];
  export type OutputTuple = [previousOwner: string, newOwner: string];
  export interface OutputObject {
    previousOwner: string;
    newOwner: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace PoolEvent {
  export type InputTuple = [
    token0: AddressLike,
    token1: AddressLike,
    pool: AddressLike
  ];
  export type OutputTuple = [token0: string, token1: string, pool: string];
  export interface OutputObject {
    token0: string;
    token1: string;
    pool: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace RenounceOwnershipFinishEvent {
  export type InputTuple = [timestamp: BigNumberish];
  export type OutputTuple = [timestamp: bigint];
  export interface OutputObject {
    timestamp: bigint;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace RenounceOwnershipStartEvent {
  export type InputTuple = [
    timestamp: BigNumberish,
    finishTimestamp: BigNumberish
  ];
  export type OutputTuple = [timestamp: bigint, finishTimestamp: bigint];
  export interface OutputObject {
    timestamp: bigint;
    finishTimestamp: bigint;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace RenounceOwnershipStopEvent {
  export type InputTuple = [timestamp: BigNumberish];
  export type OutputTuple = [timestamp: bigint];
  export interface OutputObject {
    timestamp: bigint;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace RoleAdminChangedEvent {
  export type InputTuple = [
    role: BytesLike,
    previousAdminRole: BytesLike,
    newAdminRole: BytesLike
  ];
  export type OutputTuple = [
    role: string,
    previousAdminRole: string,
    newAdminRole: string
  ];
  export interface OutputObject {
    role: string;
    previousAdminRole: string;
    newAdminRole: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace RoleGrantedEvent {
  export type InputTuple = [
    role: BytesLike,
    account: AddressLike,
    sender: AddressLike
  ];
  export type OutputTuple = [role: string, account: string, sender: string];
  export interface OutputObject {
    role: string;
    account: string;
    sender: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace RoleRevokedEvent {
  export type InputTuple = [
    role: BytesLike,
    account: AddressLike,
    sender: AddressLike
  ];
  export type OutputTuple = [role: string, account: string, sender: string];
  export interface OutputObject {
    role: string;
    account: string;
    sender: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace VaultFactoryEvent {
  export type InputTuple = [newVaultFactory: AddressLike];
  export type OutputTuple = [newVaultFactory: string];
  export interface OutputObject {
    newVaultFactory: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export interface AlgebraFactory extends BaseContract {
  connect(runner?: ContractRunner | null): AlgebraFactory;
  waitForDeployment(): Promise<this>;

  interface: AlgebraFactoryInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  CUSTOM_POOL_DEPLOYER: TypedContractMethod<[], [string], "view">;

  DEFAULT_ADMIN_ROLE: TypedContractMethod<[], [string], "view">;

  POOLS_ADMINISTRATOR_ROLE: TypedContractMethod<[], [string], "view">;

  POOL_INIT_CODE_HASH: TypedContractMethod<[], [string], "view">;

  acceptOwnership: TypedContractMethod<[], [void], "nonpayable">;

  computeCustomPoolAddress: TypedContractMethod<
    [deployer: AddressLike, token0: AddressLike, token1: AddressLike],
    [string],
    "view"
  >;

  computePoolAddress: TypedContractMethod<
    [token0: AddressLike, token1: AddressLike],
    [string],
    "view"
  >;

  createCustomPool: TypedContractMethod<
    [
      deployer: AddressLike,
      creator: AddressLike,
      tokenA: AddressLike,
      tokenB: AddressLike,
      data: BytesLike
    ],
    [string],
    "nonpayable"
  >;

  createPool: TypedContractMethod<
    [tokenA: AddressLike, tokenB: AddressLike, data: BytesLike],
    [string],
    "nonpayable"
  >;

  customPoolByPair: TypedContractMethod<
    [arg0: AddressLike, arg1: AddressLike, arg2: AddressLike],
    [string],
    "view"
  >;

  defaultCommunityFee: TypedContractMethod<[], [bigint], "view">;

  defaultConfigurationForPool: TypedContractMethod<
    [],
    [
      [bigint, bigint, bigint] & {
        communityFee: bigint;
        tickSpacing: bigint;
        fee: bigint;
      }
    ],
    "view"
  >;

  defaultFee: TypedContractMethod<[], [bigint], "view">;

  defaultPluginFactory: TypedContractMethod<[], [string], "view">;

  defaultTickspacing: TypedContractMethod<[], [bigint], "view">;

  getRoleAdmin: TypedContractMethod<[role: BytesLike], [string], "view">;

  getRoleMember: TypedContractMethod<
    [role: BytesLike, index: BigNumberish],
    [string],
    "view"
  >;

  getRoleMemberCount: TypedContractMethod<[role: BytesLike], [bigint], "view">;

  grantRole: TypedContractMethod<
    [role: BytesLike, account: AddressLike],
    [void],
    "nonpayable"
  >;

  hasRole: TypedContractMethod<
    [role: BytesLike, account: AddressLike],
    [boolean],
    "view"
  >;

  hasRoleOrOwner: TypedContractMethod<
    [role: BytesLike, account: AddressLike],
    [boolean],
    "view"
  >;

  owner: TypedContractMethod<[], [string], "view">;

  pendingOwner: TypedContractMethod<[], [string], "view">;

  poolByPair: TypedContractMethod<
    [arg0: AddressLike, arg1: AddressLike],
    [string],
    "view"
  >;

  poolDeployer: TypedContractMethod<[], [string], "view">;

  renounceOwnership: TypedContractMethod<[], [void], "nonpayable">;

  renounceOwnershipStartTimestamp: TypedContractMethod<[], [bigint], "view">;

  renounceRole: TypedContractMethod<
    [role: BytesLike, account: AddressLike],
    [void],
    "nonpayable"
  >;

  revokeRole: TypedContractMethod<
    [role: BytesLike, account: AddressLike],
    [void],
    "nonpayable"
  >;

  setDefaultCommunityFee: TypedContractMethod<
    [newDefaultCommunityFee: BigNumberish],
    [void],
    "nonpayable"
  >;

  setDefaultFee: TypedContractMethod<
    [newDefaultFee: BigNumberish],
    [void],
    "nonpayable"
  >;

  setDefaultPluginFactory: TypedContractMethod<
    [newDefaultPluginFactory: AddressLike],
    [void],
    "nonpayable"
  >;

  setDefaultTickspacing: TypedContractMethod<
    [newDefaultTickspacing: BigNumberish],
    [void],
    "nonpayable"
  >;

  setVaultFactory: TypedContractMethod<
    [newVaultFactory: AddressLike],
    [void],
    "nonpayable"
  >;

  startRenounceOwnership: TypedContractMethod<[], [void], "nonpayable">;

  stopRenounceOwnership: TypedContractMethod<[], [void], "nonpayable">;

  supportsInterface: TypedContractMethod<
    [interfaceId: BytesLike],
    [boolean],
    "view"
  >;

  transferOwnership: TypedContractMethod<
    [newOwner: AddressLike],
    [void],
    "nonpayable"
  >;

  vaultFactory: TypedContractMethod<[], [string], "view">;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "CUSTOM_POOL_DEPLOYER"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "DEFAULT_ADMIN_ROLE"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "POOLS_ADMINISTRATOR_ROLE"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "POOL_INIT_CODE_HASH"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "acceptOwnership"
  ): TypedContractMethod<[], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "computeCustomPoolAddress"
  ): TypedContractMethod<
    [deployer: AddressLike, token0: AddressLike, token1: AddressLike],
    [string],
    "view"
  >;
  getFunction(
    nameOrSignature: "computePoolAddress"
  ): TypedContractMethod<
    [token0: AddressLike, token1: AddressLike],
    [string],
    "view"
  >;
  getFunction(
    nameOrSignature: "createCustomPool"
  ): TypedContractMethod<
    [
      deployer: AddressLike,
      creator: AddressLike,
      tokenA: AddressLike,
      tokenB: AddressLike,
      data: BytesLike
    ],
    [string],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "createPool"
  ): TypedContractMethod<
    [tokenA: AddressLike, tokenB: AddressLike, data: BytesLike],
    [string],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "customPoolByPair"
  ): TypedContractMethod<
    [arg0: AddressLike, arg1: AddressLike, arg2: AddressLike],
    [string],
    "view"
  >;
  getFunction(
    nameOrSignature: "defaultCommunityFee"
  ): TypedContractMethod<[], [bigint], "view">;
  getFunction(
    nameOrSignature: "defaultConfigurationForPool"
  ): TypedContractMethod<
    [],
    [
      [bigint, bigint, bigint] & {
        communityFee: bigint;
        tickSpacing: bigint;
        fee: bigint;
      }
    ],
    "view"
  >;
  getFunction(
    nameOrSignature: "defaultFee"
  ): TypedContractMethod<[], [bigint], "view">;
  getFunction(
    nameOrSignature: "defaultPluginFactory"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "defaultTickspacing"
  ): TypedContractMethod<[], [bigint], "view">;
  getFunction(
    nameOrSignature: "getRoleAdmin"
  ): TypedContractMethod<[role: BytesLike], [string], "view">;
  getFunction(
    nameOrSignature: "getRoleMember"
  ): TypedContractMethod<
    [role: BytesLike, index: BigNumberish],
    [string],
    "view"
  >;
  getFunction(
    nameOrSignature: "getRoleMemberCount"
  ): TypedContractMethod<[role: BytesLike], [bigint], "view">;
  getFunction(
    nameOrSignature: "grantRole"
  ): TypedContractMethod<
    [role: BytesLike, account: AddressLike],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "hasRole"
  ): TypedContractMethod<
    [role: BytesLike, account: AddressLike],
    [boolean],
    "view"
  >;
  getFunction(
    nameOrSignature: "hasRoleOrOwner"
  ): TypedContractMethod<
    [role: BytesLike, account: AddressLike],
    [boolean],
    "view"
  >;
  getFunction(
    nameOrSignature: "owner"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "pendingOwner"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "poolByPair"
  ): TypedContractMethod<
    [arg0: AddressLike, arg1: AddressLike],
    [string],
    "view"
  >;
  getFunction(
    nameOrSignature: "poolDeployer"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "renounceOwnership"
  ): TypedContractMethod<[], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "renounceOwnershipStartTimestamp"
  ): TypedContractMethod<[], [bigint], "view">;
  getFunction(
    nameOrSignature: "renounceRole"
  ): TypedContractMethod<
    [role: BytesLike, account: AddressLike],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "revokeRole"
  ): TypedContractMethod<
    [role: BytesLike, account: AddressLike],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "setDefaultCommunityFee"
  ): TypedContractMethod<
    [newDefaultCommunityFee: BigNumberish],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "setDefaultFee"
  ): TypedContractMethod<[newDefaultFee: BigNumberish], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "setDefaultPluginFactory"
  ): TypedContractMethod<
    [newDefaultPluginFactory: AddressLike],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "setDefaultTickspacing"
  ): TypedContractMethod<
    [newDefaultTickspacing: BigNumberish],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "setVaultFactory"
  ): TypedContractMethod<[newVaultFactory: AddressLike], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "startRenounceOwnership"
  ): TypedContractMethod<[], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "stopRenounceOwnership"
  ): TypedContractMethod<[], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "supportsInterface"
  ): TypedContractMethod<[interfaceId: BytesLike], [boolean], "view">;
  getFunction(
    nameOrSignature: "transferOwnership"
  ): TypedContractMethod<[newOwner: AddressLike], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "vaultFactory"
  ): TypedContractMethod<[], [string], "view">;

  getEvent(
    key: "CustomPool"
  ): TypedContractEvent<
    CustomPoolEvent.InputTuple,
    CustomPoolEvent.OutputTuple,
    CustomPoolEvent.OutputObject
  >;
  getEvent(
    key: "DefaultCommunityFee"
  ): TypedContractEvent<
    DefaultCommunityFeeEvent.InputTuple,
    DefaultCommunityFeeEvent.OutputTuple,
    DefaultCommunityFeeEvent.OutputObject
  >;
  getEvent(
    key: "DefaultFee"
  ): TypedContractEvent<
    DefaultFeeEvent.InputTuple,
    DefaultFeeEvent.OutputTuple,
    DefaultFeeEvent.OutputObject
  >;
  getEvent(
    key: "DefaultPluginFactory"
  ): TypedContractEvent<
    DefaultPluginFactoryEvent.InputTuple,
    DefaultPluginFactoryEvent.OutputTuple,
    DefaultPluginFactoryEvent.OutputObject
  >;
  getEvent(
    key: "DefaultTickspacing"
  ): TypedContractEvent<
    DefaultTickspacingEvent.InputTuple,
    DefaultTickspacingEvent.OutputTuple,
    DefaultTickspacingEvent.OutputObject
  >;
  getEvent(
    key: "OwnershipTransferStarted"
  ): TypedContractEvent<
    OwnershipTransferStartedEvent.InputTuple,
    OwnershipTransferStartedEvent.OutputTuple,
    OwnershipTransferStartedEvent.OutputObject
  >;
  getEvent(
    key: "OwnershipTransferred"
  ): TypedContractEvent<
    OwnershipTransferredEvent.InputTuple,
    OwnershipTransferredEvent.OutputTuple,
    OwnershipTransferredEvent.OutputObject
  >;
  getEvent(
    key: "Pool"
  ): TypedContractEvent<
    PoolEvent.InputTuple,
    PoolEvent.OutputTuple,
    PoolEvent.OutputObject
  >;
  getEvent(
    key: "RenounceOwnershipFinish"
  ): TypedContractEvent<
    RenounceOwnershipFinishEvent.InputTuple,
    RenounceOwnershipFinishEvent.OutputTuple,
    RenounceOwnershipFinishEvent.OutputObject
  >;
  getEvent(
    key: "RenounceOwnershipStart"
  ): TypedContractEvent<
    RenounceOwnershipStartEvent.InputTuple,
    RenounceOwnershipStartEvent.OutputTuple,
    RenounceOwnershipStartEvent.OutputObject
  >;
  getEvent(
    key: "RenounceOwnershipStop"
  ): TypedContractEvent<
    RenounceOwnershipStopEvent.InputTuple,
    RenounceOwnershipStopEvent.OutputTuple,
    RenounceOwnershipStopEvent.OutputObject
  >;
  getEvent(
    key: "RoleAdminChanged"
  ): TypedContractEvent<
    RoleAdminChangedEvent.InputTuple,
    RoleAdminChangedEvent.OutputTuple,
    RoleAdminChangedEvent.OutputObject
  >;
  getEvent(
    key: "RoleGranted"
  ): TypedContractEvent<
    RoleGrantedEvent.InputTuple,
    RoleGrantedEvent.OutputTuple,
    RoleGrantedEvent.OutputObject
  >;
  getEvent(
    key: "RoleRevoked"
  ): TypedContractEvent<
    RoleRevokedEvent.InputTuple,
    RoleRevokedEvent.OutputTuple,
    RoleRevokedEvent.OutputObject
  >;
  getEvent(
    key: "VaultFactory"
  ): TypedContractEvent<
    VaultFactoryEvent.InputTuple,
    VaultFactoryEvent.OutputTuple,
    VaultFactoryEvent.OutputObject
  >;

  filters: {
    "CustomPool(address,address,address,address)": TypedContractEvent<
      CustomPoolEvent.InputTuple,
      CustomPoolEvent.OutputTuple,
      CustomPoolEvent.OutputObject
    >;
    CustomPool: TypedContractEvent<
      CustomPoolEvent.InputTuple,
      CustomPoolEvent.OutputTuple,
      CustomPoolEvent.OutputObject
    >;

    "DefaultCommunityFee(uint16)": TypedContractEvent<
      DefaultCommunityFeeEvent.InputTuple,
      DefaultCommunityFeeEvent.OutputTuple,
      DefaultCommunityFeeEvent.OutputObject
    >;
    DefaultCommunityFee: TypedContractEvent<
      DefaultCommunityFeeEvent.InputTuple,
      DefaultCommunityFeeEvent.OutputTuple,
      DefaultCommunityFeeEvent.OutputObject
    >;

    "DefaultFee(uint16)": TypedContractEvent<
      DefaultFeeEvent.InputTuple,
      DefaultFeeEvent.OutputTuple,
      DefaultFeeEvent.OutputObject
    >;
    DefaultFee: TypedContractEvent<
      DefaultFeeEvent.InputTuple,
      DefaultFeeEvent.OutputTuple,
      DefaultFeeEvent.OutputObject
    >;

    "DefaultPluginFactory(address)": TypedContractEvent<
      DefaultPluginFactoryEvent.InputTuple,
      DefaultPluginFactoryEvent.OutputTuple,
      DefaultPluginFactoryEvent.OutputObject
    >;
    DefaultPluginFactory: TypedContractEvent<
      DefaultPluginFactoryEvent.InputTuple,
      DefaultPluginFactoryEvent.OutputTuple,
      DefaultPluginFactoryEvent.OutputObject
    >;

    "DefaultTickspacing(int24)": TypedContractEvent<
      DefaultTickspacingEvent.InputTuple,
      DefaultTickspacingEvent.OutputTuple,
      DefaultTickspacingEvent.OutputObject
    >;
    DefaultTickspacing: TypedContractEvent<
      DefaultTickspacingEvent.InputTuple,
      DefaultTickspacingEvent.OutputTuple,
      DefaultTickspacingEvent.OutputObject
    >;

    "OwnershipTransferStarted(address,address)": TypedContractEvent<
      OwnershipTransferStartedEvent.InputTuple,
      OwnershipTransferStartedEvent.OutputTuple,
      OwnershipTransferStartedEvent.OutputObject
    >;
    OwnershipTransferStarted: TypedContractEvent<
      OwnershipTransferStartedEvent.InputTuple,
      OwnershipTransferStartedEvent.OutputTuple,
      OwnershipTransferStartedEvent.OutputObject
    >;

    "OwnershipTransferred(address,address)": TypedContractEvent<
      OwnershipTransferredEvent.InputTuple,
      OwnershipTransferredEvent.OutputTuple,
      OwnershipTransferredEvent.OutputObject
    >;
    OwnershipTransferred: TypedContractEvent<
      OwnershipTransferredEvent.InputTuple,
      OwnershipTransferredEvent.OutputTuple,
      OwnershipTransferredEvent.OutputObject
    >;

    "Pool(address,address,address)": TypedContractEvent<
      PoolEvent.InputTuple,
      PoolEvent.OutputTuple,
      PoolEvent.OutputObject
    >;
    Pool: TypedContractEvent<
      PoolEvent.InputTuple,
      PoolEvent.OutputTuple,
      PoolEvent.OutputObject
    >;

    "RenounceOwnershipFinish(uint256)": TypedContractEvent<
      RenounceOwnershipFinishEvent.InputTuple,
      RenounceOwnershipFinishEvent.OutputTuple,
      RenounceOwnershipFinishEvent.OutputObject
    >;
    RenounceOwnershipFinish: TypedContractEvent<
      RenounceOwnershipFinishEvent.InputTuple,
      RenounceOwnershipFinishEvent.OutputTuple,
      RenounceOwnershipFinishEvent.OutputObject
    >;

    "RenounceOwnershipStart(uint256,uint256)": TypedContractEvent<
      RenounceOwnershipStartEvent.InputTuple,
      RenounceOwnershipStartEvent.OutputTuple,
      RenounceOwnershipStartEvent.OutputObject
    >;
    RenounceOwnershipStart: TypedContractEvent<
      RenounceOwnershipStartEvent.InputTuple,
      RenounceOwnershipStartEvent.OutputTuple,
      RenounceOwnershipStartEvent.OutputObject
    >;

    "RenounceOwnershipStop(uint256)": TypedContractEvent<
      RenounceOwnershipStopEvent.InputTuple,
      RenounceOwnershipStopEvent.OutputTuple,
      RenounceOwnershipStopEvent.OutputObject
    >;
    RenounceOwnershipStop: TypedContractEvent<
      RenounceOwnershipStopEvent.InputTuple,
      RenounceOwnershipStopEvent.OutputTuple,
      RenounceOwnershipStopEvent.OutputObject
    >;

    "RoleAdminChanged(bytes32,bytes32,bytes32)": TypedContractEvent<
      RoleAdminChangedEvent.InputTuple,
      RoleAdminChangedEvent.OutputTuple,
      RoleAdminChangedEvent.OutputObject
    >;
    RoleAdminChanged: TypedContractEvent<
      RoleAdminChangedEvent.InputTuple,
      RoleAdminChangedEvent.OutputTuple,
      RoleAdminChangedEvent.OutputObject
    >;

    "RoleGranted(bytes32,address,address)": TypedContractEvent<
      RoleGrantedEvent.InputTuple,
      RoleGrantedEvent.OutputTuple,
      RoleGrantedEvent.OutputObject
    >;
    RoleGranted: TypedContractEvent<
      RoleGrantedEvent.InputTuple,
      RoleGrantedEvent.OutputTuple,
      RoleGrantedEvent.OutputObject
    >;

    "RoleRevoked(bytes32,address,address)": TypedContractEvent<
      RoleRevokedEvent.InputTuple,
      RoleRevokedEvent.OutputTuple,
      RoleRevokedEvent.OutputObject
    >;
    RoleRevoked: TypedContractEvent<
      RoleRevokedEvent.InputTuple,
      RoleRevokedEvent.OutputTuple,
      RoleRevokedEvent.OutputObject
    >;

    "VaultFactory(address)": TypedContractEvent<
      VaultFactoryEvent.InputTuple,
      VaultFactoryEvent.OutputTuple,
      VaultFactoryEvent.OutputObject
    >;
    VaultFactory: TypedContractEvent<
      VaultFactoryEvent.InputTuple,
      VaultFactoryEvent.OutputTuple,
      VaultFactoryEvent.OutputObject
    >;
  };
}
