// @cryptoalgebra/integral-core/typechain/contracts/AlgebraPoolDeployer.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  AddressLike,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedListener,
  TypedContractMethod,
} from "../common";

export interface AlgebraPoolDeployerInterface extends Interface {
  getFunction(
    nameOrSignature: "deploy" | "getDeployParameters"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "deploy",
    values: [AddressLike, AddressLike, AddressLike, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "getDeployParameters",
    values?: undefined
  ): string;

  decodeFunctionResult(functionFragment: "deploy", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "getDeployParameters",
    data: BytesLike
  ): Result;
}

export interface AlgebraPoolDeployer extends BaseContract {
  connect(runner?: ContractRunner | null): AlgebraPoolDeployer;
  waitForDeployment(): Promise<this>;

  interface: AlgebraPoolDeployerInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  deploy: TypedContractMethod<
    [
      plugin: AddressLike,
      token0: AddressLike,
      token1: AddressLike,
      deployer: AddressLike
    ],
    [string],
    "nonpayable"
  >;

  getDeployParameters: TypedContractMethod<
    [],
    [
      [string, string, string, string] & {
        _plugin: string;
        _factory: string;
        _token0: string;
        _token1: string;
      }
    ],
    "view"
  >;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "deploy"
  ): TypedContractMethod<
    [
      plugin: AddressLike,
      token0: AddressLike,
      token1: AddressLike,
      deployer: AddressLike
    ],
    [string],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "getDeployParameters"
  ): TypedContractMethod<
    [],
    [
      [string, string, string, string] & {
        _plugin: string;
        _factory: string;
        _token0: string;
        _token1: string;
      }
    ],
    "view"
  >;

  filters: {};
}
