// @cryptoalgebra/integral-core/typechain/contracts/interfaces/vault/IAlgebraCommunityVault.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumberish,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  EventFragment,
  AddressLike,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedLogDescription,
  TypedListener,
  TypedContractMethod,
} from "../../../common";

export declare namespace IAlgebraCommunityVault {
  export type WithdrawTokensParamsStruct = {
    token: AddressLike;
    amount: BigNumberish;
  };

  export type WithdrawTokensParamsStructOutput = [
    token: string,
    amount: bigint
  ] & { token: string; amount: bigint };
}

export interface IAlgebraCommunityVaultInterface extends Interface {
  getFunction(
    nameOrSignature:
      | "acceptAlgebraFeeChangeProposal"
      | "acceptAlgebraFeeManagerRole"
      | "cancelAlgebraFeeChangeProposal"
      | "changeAlgebraFeeReceiver"
      | "changeCommunityFeeReceiver"
      | "proposeAlgebraFeeChange"
      | "transferAlgebraFeeManagerRole"
      | "withdraw"
      | "withdrawTokens"
  ): FunctionFragment;

  getEvent(
    nameOrSignatureOrTopic:
      | "AlgebraFee"
      | "AlgebraFeeManager"
      | "AlgebraFeeProposal"
      | "AlgebraFeeReceiver"
      | "AlgebraTokensWithdrawal"
      | "CancelAlgebraFeeProposal"
      | "CommunityFeeReceiver"
      | "PendingAlgebraFeeManager"
      | "TokensWithdrawal"
  ): EventFragment;

  encodeFunctionData(
    functionFragment: "acceptAlgebraFeeChangeProposal",
    values: [BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "acceptAlgebraFeeManagerRole",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "cancelAlgebraFeeChangeProposal",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "changeAlgebraFeeReceiver",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "changeCommunityFeeReceiver",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "proposeAlgebraFeeChange",
    values: [BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "transferAlgebraFeeManagerRole",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "withdraw",
    values: [AddressLike, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "withdrawTokens",
    values: [IAlgebraCommunityVault.WithdrawTokensParamsStruct[]]
  ): string;

  decodeFunctionResult(
    functionFragment: "acceptAlgebraFeeChangeProposal",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "acceptAlgebraFeeManagerRole",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "cancelAlgebraFeeChangeProposal",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "changeAlgebraFeeReceiver",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "changeCommunityFeeReceiver",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "proposeAlgebraFeeChange",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "transferAlgebraFeeManagerRole",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "withdraw", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "withdrawTokens",
    data: BytesLike
  ): Result;
}

export namespace AlgebraFeeEvent {
  export type InputTuple = [newAlgebraFee: BigNumberish];
  export type OutputTuple = [newAlgebraFee: bigint];
  export interface OutputObject {
    newAlgebraFee: bigint;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace AlgebraFeeManagerEvent {
  export type InputTuple = [newAlgebraFeeManager: AddressLike];
  export type OutputTuple = [newAlgebraFeeManager: string];
  export interface OutputObject {
    newAlgebraFeeManager: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace AlgebraFeeProposalEvent {
  export type InputTuple = [proposedNewAlgebraFee: BigNumberish];
  export type OutputTuple = [proposedNewAlgebraFee: bigint];
  export interface OutputObject {
    proposedNewAlgebraFee: bigint;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace AlgebraFeeReceiverEvent {
  export type InputTuple = [newAlgebraFeeReceiver: AddressLike];
  export type OutputTuple = [newAlgebraFeeReceiver: string];
  export interface OutputObject {
    newAlgebraFeeReceiver: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace AlgebraTokensWithdrawalEvent {
  export type InputTuple = [
    token: AddressLike,
    to: AddressLike,
    amount: BigNumberish
  ];
  export type OutputTuple = [token: string, to: string, amount: bigint];
  export interface OutputObject {
    token: string;
    to: string;
    amount: bigint;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace CancelAlgebraFeeProposalEvent {
  export type InputTuple = [];
  export type OutputTuple = [];
  export interface OutputObject {}
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace CommunityFeeReceiverEvent {
  export type InputTuple = [newCommunityFeeReceiver: AddressLike];
  export type OutputTuple = [newCommunityFeeReceiver: string];
  export interface OutputObject {
    newCommunityFeeReceiver: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace PendingAlgebraFeeManagerEvent {
  export type InputTuple = [pendingAlgebraFeeManager: AddressLike];
  export type OutputTuple = [pendingAlgebraFeeManager: string];
  export interface OutputObject {
    pendingAlgebraFeeManager: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace TokensWithdrawalEvent {
  export type InputTuple = [
    token: AddressLike,
    to: AddressLike,
    amount: BigNumberish
  ];
  export type OutputTuple = [token: string, to: string, amount: bigint];
  export interface OutputObject {
    token: string;
    to: string;
    amount: bigint;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export interface IAlgebraCommunityVault extends BaseContract {
  connect(runner?: ContractRunner | null): IAlgebraCommunityVault;
  waitForDeployment(): Promise<this>;

  interface: IAlgebraCommunityVaultInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  acceptAlgebraFeeChangeProposal: TypedContractMethod<
    [newAlgebraFee: BigNumberish],
    [void],
    "nonpayable"
  >;

  acceptAlgebraFeeManagerRole: TypedContractMethod<[], [void], "nonpayable">;

  cancelAlgebraFeeChangeProposal: TypedContractMethod<[], [void], "nonpayable">;

  changeAlgebraFeeReceiver: TypedContractMethod<
    [newAlgebraFeeReceiver: AddressLike],
    [void],
    "nonpayable"
  >;

  changeCommunityFeeReceiver: TypedContractMethod<
    [newCommunityFeeReceiver: AddressLike],
    [void],
    "nonpayable"
  >;

  proposeAlgebraFeeChange: TypedContractMethod<
    [newAlgebraFee: BigNumberish],
    [void],
    "nonpayable"
  >;

  transferAlgebraFeeManagerRole: TypedContractMethod<
    [_newAlgebraFeeManager: AddressLike],
    [void],
    "nonpayable"
  >;

  withdraw: TypedContractMethod<
    [token: AddressLike, amount: BigNumberish],
    [void],
    "nonpayable"
  >;

  withdrawTokens: TypedContractMethod<
    [params: IAlgebraCommunityVault.WithdrawTokensParamsStruct[]],
    [void],
    "nonpayable"
  >;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "acceptAlgebraFeeChangeProposal"
  ): TypedContractMethod<[newAlgebraFee: BigNumberish], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "acceptAlgebraFeeManagerRole"
  ): TypedContractMethod<[], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "cancelAlgebraFeeChangeProposal"
  ): TypedContractMethod<[], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "changeAlgebraFeeReceiver"
  ): TypedContractMethod<
    [newAlgebraFeeReceiver: AddressLike],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "changeCommunityFeeReceiver"
  ): TypedContractMethod<
    [newCommunityFeeReceiver: AddressLike],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "proposeAlgebraFeeChange"
  ): TypedContractMethod<[newAlgebraFee: BigNumberish], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "transferAlgebraFeeManagerRole"
  ): TypedContractMethod<
    [_newAlgebraFeeManager: AddressLike],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "withdraw"
  ): TypedContractMethod<
    [token: AddressLike, amount: BigNumberish],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "withdrawTokens"
  ): TypedContractMethod<
    [params: IAlgebraCommunityVault.WithdrawTokensParamsStruct[]],
    [void],
    "nonpayable"
  >;

  getEvent(
    key: "AlgebraFee"
  ): TypedContractEvent<
    AlgebraFeeEvent.InputTuple,
    AlgebraFeeEvent.OutputTuple,
    AlgebraFeeEvent.OutputObject
  >;
  getEvent(
    key: "AlgebraFeeManager"
  ): TypedContractEvent<
    AlgebraFeeManagerEvent.InputTuple,
    AlgebraFeeManagerEvent.OutputTuple,
    AlgebraFeeManagerEvent.OutputObject
  >;
  getEvent(
    key: "AlgebraFeeProposal"
  ): TypedContractEvent<
    AlgebraFeeProposalEvent.InputTuple,
    AlgebraFeeProposalEvent.OutputTuple,
    AlgebraFeeProposalEvent.OutputObject
  >;
  getEvent(
    key: "AlgebraFeeReceiver"
  ): TypedContractEvent<
    AlgebraFeeReceiverEvent.InputTuple,
    AlgebraFeeReceiverEvent.OutputTuple,
    AlgebraFeeReceiverEvent.OutputObject
  >;
  getEvent(
    key: "AlgebraTokensWithdrawal"
  ): TypedContractEvent<
    AlgebraTokensWithdrawalEvent.InputTuple,
    AlgebraTokensWithdrawalEvent.OutputTuple,
    AlgebraTokensWithdrawalEvent.OutputObject
  >;
  getEvent(
    key: "CancelAlgebraFeeProposal"
  ): TypedContractEvent<
    CancelAlgebraFeeProposalEvent.InputTuple,
    CancelAlgebraFeeProposalEvent.OutputTuple,
    CancelAlgebraFeeProposalEvent.OutputObject
  >;
  getEvent(
    key: "CommunityFeeReceiver"
  ): TypedContractEvent<
    CommunityFeeReceiverEvent.InputTuple,
    CommunityFeeReceiverEvent.OutputTuple,
    CommunityFeeReceiverEvent.OutputObject
  >;
  getEvent(
    key: "PendingAlgebraFeeManager"
  ): TypedContractEvent<
    PendingAlgebraFeeManagerEvent.InputTuple,
    PendingAlgebraFeeManagerEvent.OutputTuple,
    PendingAlgebraFeeManagerEvent.OutputObject
  >;
  getEvent(
    key: "TokensWithdrawal"
  ): TypedContractEvent<
    TokensWithdrawalEvent.InputTuple,
    TokensWithdrawalEvent.OutputTuple,
    TokensWithdrawalEvent.OutputObject
  >;

  filters: {
    "AlgebraFee(uint16)": TypedContractEvent<
      AlgebraFeeEvent.InputTuple,
      AlgebraFeeEvent.OutputTuple,
      AlgebraFeeEvent.OutputObject
    >;
    AlgebraFee: TypedContractEvent<
      AlgebraFeeEvent.InputTuple,
      AlgebraFeeEvent.OutputTuple,
      AlgebraFeeEvent.OutputObject
    >;

    "AlgebraFeeManager(address)": TypedContractEvent<
      AlgebraFeeManagerEvent.InputTuple,
      AlgebraFeeManagerEvent.OutputTuple,
      AlgebraFeeManagerEvent.OutputObject
    >;
    AlgebraFeeManager: TypedContractEvent<
      AlgebraFeeManagerEvent.InputTuple,
      AlgebraFeeManagerEvent.OutputTuple,
      AlgebraFeeManagerEvent.OutputObject
    >;

    "AlgebraFeeProposal(uint16)": TypedContractEvent<
      AlgebraFeeProposalEvent.InputTuple,
      AlgebraFeeProposalEvent.OutputTuple,
      AlgebraFeeProposalEvent.OutputObject
    >;
    AlgebraFeeProposal: TypedContractEvent<
      AlgebraFeeProposalEvent.InputTuple,
      AlgebraFeeProposalEvent.OutputTuple,
      AlgebraFeeProposalEvent.OutputObject
    >;

    "AlgebraFeeReceiver(address)": TypedContractEvent<
      AlgebraFeeReceiverEvent.InputTuple,
      AlgebraFeeReceiverEvent.OutputTuple,
      AlgebraFeeReceiverEvent.OutputObject
    >;
    AlgebraFeeReceiver: TypedContractEvent<
      AlgebraFeeReceiverEvent.InputTuple,
      AlgebraFeeReceiverEvent.OutputTuple,
      AlgebraFeeReceiverEvent.OutputObject
    >;

    "AlgebraTokensWithdrawal(address,address,uint256)": TypedContractEvent<
      AlgebraTokensWithdrawalEvent.InputTuple,
      AlgebraTokensWithdrawalEvent.OutputTuple,
      AlgebraTokensWithdrawalEvent.OutputObject
    >;
    AlgebraTokensWithdrawal: TypedContractEvent<
      AlgebraTokensWithdrawalEvent.InputTuple,
      AlgebraTokensWithdrawalEvent.OutputTuple,
      AlgebraTokensWithdrawalEvent.OutputObject
    >;

    "CancelAlgebraFeeProposal()": TypedContractEvent<
      CancelAlgebraFeeProposalEvent.InputTuple,
      CancelAlgebraFeeProposalEvent.OutputTuple,
      CancelAlgebraFeeProposalEvent.OutputObject
    >;
    CancelAlgebraFeeProposal: TypedContractEvent<
      CancelAlgebraFeeProposalEvent.InputTuple,
      CancelAlgebraFeeProposalEvent.OutputTuple,
      CancelAlgebraFeeProposalEvent.OutputObject
    >;

    "CommunityFeeReceiver(address)": TypedContractEvent<
      CommunityFeeReceiverEvent.InputTuple,
      CommunityFeeReceiverEvent.OutputTuple,
      CommunityFeeReceiverEvent.OutputObject
    >;
    CommunityFeeReceiver: TypedContractEvent<
      CommunityFeeReceiverEvent.InputTuple,
      CommunityFeeReceiverEvent.OutputTuple,
      CommunityFeeReceiverEvent.OutputObject
    >;

    "PendingAlgebraFeeManager(address)": TypedContractEvent<
      PendingAlgebraFeeManagerEvent.InputTuple,
      PendingAlgebraFeeManagerEvent.OutputTuple,
      PendingAlgebraFeeManagerEvent.OutputObject
    >;
    PendingAlgebraFeeManager: TypedContractEvent<
      PendingAlgebraFeeManagerEvent.InputTuple,
      PendingAlgebraFeeManagerEvent.OutputTuple,
      PendingAlgebraFeeManagerEvent.OutputObject
    >;

    "TokensWithdrawal(address,address,uint256)": TypedContractEvent<
      TokensWithdrawalEvent.InputTuple,
      TokensWithdrawalEvent.OutputTuple,
      TokensWithdrawalEvent.OutputObject
    >;
    TokensWithdrawal: TypedContractEvent<
      TokensWithdrawalEvent.InputTuple,
      TokensWithdrawalEvent.OutputTuple,
      TokensWithdrawalEvent.OutputObject
    >;
  };
}
