// @cryptoalgebra/integral-core/typechain/contracts/interfaces/vault/IAlgebraVaultFactory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  AddressLike,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedListener,
  TypedContractMethod,
} from "../../../common";

export interface IAlgebraVaultFactoryInterface extends Interface {
  getFunction(
    nameOrSignature: "createVaultForPool" | "getVaultForPool"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "createVaultForPool",
    values: [AddressLike, AddressLike, AddressLike, AddressLike, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "getVaultForPool",
    values: [AddressLike]
  ): string;

  decodeFunctionResult(
    functionFragment: "createVaultForPool",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getVaultForPool",
    data: BytesLike
  ): Result;
}

export interface IAlgebraVaultFactory extends BaseContract {
  connect(runner?: ContractRunner | null): IAlgebraVaultFactory;
  waitForDeployment(): Promise<this>;

  interface: IAlgebraVaultFactoryInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  createVaultForPool: TypedContractMethod<
    [
      pool: AddressLike,
      creator: AddressLike,
      deployer: AddressLike,
      token0: AddressLike,
      token1: AddressLike
    ],
    [string],
    "nonpayable"
  >;

  getVaultForPool: TypedContractMethod<[pool: AddressLike], [string], "view">;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "createVaultForPool"
  ): TypedContractMethod<
    [
      pool: AddressLike,
      creator: AddressLike,
      deployer: AddressLike,
      token0: AddressLike,
      token1: AddressLike
    ],
    [string],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "getVaultForPool"
  ): TypedContractMethod<[pool: AddressLike], [string], "view">;

  filters: {};
}
