// @cryptoalgebra/integral-core/typechain/factories/contracts/interfaces/vault/IAlgebraCommunityVault__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Interface, type ContractRunner } from "ethers";
import type {
  IAlgebraCommunityVault,
  IAlgebraCommunityVaultInterface,
} from "../../../../contracts/interfaces/vault/IAlgebraCommunityVault";

const _abi = [
  {
    anonymous: false,
    inputs: [
      {
        indexed: false,
        internalType: "uint16",
        name: "newAlgebraFee",
        type: "uint16",
      },
    ],
    name: "AlgebraFee",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: false,
        internalType: "address",
        name: "newAlgebraFeeManager",
        type: "address",
      },
    ],
    name: "AlgebraFeeManager",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: false,
        internalType: "uint16",
        name: "proposedNewAlgebraFee",
        type: "uint16",
      },
    ],
    name: "AlgebraFeeProposal",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: false,
        internalType: "address",
        name: "newAlgebraFeeReceiver",
        type: "address",
      },
    ],
    name: "AlgebraFeeReceiver",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "token",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "to",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "amount",
        type: "uint256",
      },
    ],
    name: "AlgebraTokensWithdrawal",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [],
    name: "CancelAlgebraFeeProposal",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: false,
        internalType: "address",
        name: "newCommunityFeeReceiver",
        type: "address",
      },
    ],
    name: "CommunityFeeReceiver",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: false,
        internalType: "address",
        name: "pendingAlgebraFeeManager",
        type: "address",
      },
    ],
    name: "PendingAlgebraFeeManager",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "token",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "to",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "amount",
        type: "uint256",
      },
    ],
    name: "TokensWithdrawal",
    type: "event",
  },
  {
    inputs: [
      {
        internalType: "uint16",
        name: "newAlgebraFee",
        type: "uint16",
      },
    ],
    name: "acceptAlgebraFeeChangeProposal",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [],
    name: "acceptAlgebraFeeManagerRole",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [],
    name: "cancelAlgebraFeeChangeProposal",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "newAlgebraFeeReceiver",
        type: "address",
      },
    ],
    name: "changeAlgebraFeeReceiver",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "newCommunityFeeReceiver",
        type: "address",
      },
    ],
    name: "changeCommunityFeeReceiver",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint16",
        name: "newAlgebraFee",
        type: "uint16",
      },
    ],
    name: "proposeAlgebraFeeChange",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "_newAlgebraFeeManager",
        type: "address",
      },
    ],
    name: "transferAlgebraFeeManagerRole",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "token",
        type: "address",
      },
      {
        internalType: "uint256",
        name: "amount",
        type: "uint256",
      },
    ],
    name: "withdraw",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        components: [
          {
            internalType: "address",
            name: "token",
            type: "address",
          },
          {
            internalType: "uint256",
            name: "amount",
            type: "uint256",
          },
        ],
        internalType: "struct IAlgebraCommunityVault.WithdrawTokensParams[]",
        name: "params",
        type: "tuple[]",
      },
    ],
    name: "withdrawTokens",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
] as const;

export class IAlgebraCommunityVault__factory {
  static readonly abi = _abi;
  static createInterface(): IAlgebraCommunityVaultInterface {
    return new Interface(_abi) as IAlgebraCommunityVaultInterface;
  }
  static connect(
    address: string,
    runner?: ContractRunner | null
  ): IAlgebraCommunityVault {
    return new Contract(
      address,
      _abi,
      runner
    ) as unknown as IAlgebraCommunityVault;
  }
}
