// @cryptoalgebra/integral-core/dist/src/core/typechain/factories/contracts/interfaces/pool/IAlgebraPoolErrors__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IAlgebraPoolErrors__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "alreadyInitialized",
        type: "error",
    },
    {
        inputs: [],
        name: "arithmeticError",
        type: "error",
    },
    {
        inputs: [],
        name: "bottomTickLowerThanMIN",
        type: "error",
    },
    {
        inputs: [],
        name: "dynamicFeeActive",
        type: "error",
    },
    {
        inputs: [],
        name: "dynamicFeeDisabled",
        type: "error",
    },
    {
        inputs: [],
        name: "flashInsufficientPaid0",
        type: "error",
    },
    {
        inputs: [],
        name: "flashInsufficientPaid1",
        type: "error",
    },
    {
        inputs: [],
        name: "incorrectPluginFee",
        type: "error",
    },
    {
        inputs: [],
        name: "insufficientInputAmount",
        type: "error",
    },
    {
        inputs: [],
        name: "invalidAmountRequired",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "bytes4",
                name: "expectedSelector",
                type: "bytes4",
            },
        ],
        name: "invalidHookResponse",
        type: "error",
    },
    {
        inputs: [],
        name: "invalidLimitSqrtPrice",
        type: "error",
    },
    {
        inputs: [],
        name: "invalidNewCommunityFee",
        type: "error",
    },
    {
        inputs: [],
        name: "invalidNewTickSpacing",
        type: "error",
    },
    {
        inputs: [],
        name: "liquidityAdd",
        type: "error",
    },
    {
        inputs: [],
        name: "liquidityOverflow",
        type: "error",
    },
    {
        inputs: [],
        name: "liquiditySub",
        type: "error",
    },
    {
        inputs: [],
        name: "locked",
        type: "error",
    },
    {
        inputs: [],
        name: "notAllowed",
        type: "error",
    },
    {
        inputs: [],
        name: "notInitialized",
        type: "error",
    },
    {
        inputs: [],
        name: "pluginIsNotConnected",
        type: "error",
    },
    {
        inputs: [],
        name: "priceOutOfRange",
        type: "error",
    },
    {
        inputs: [],
        name: "tickInvalidLinks",
        type: "error",
    },
    {
        inputs: [],
        name: "tickIsNotInitialized",
        type: "error",
    },
    {
        inputs: [],
        name: "tickIsNotSpaced",
        type: "error",
    },
    {
        inputs: [],
        name: "tickOutOfRange",
        type: "error",
    },
    {
        inputs: [],
        name: "topTickAboveMAX",
        type: "error",
    },
    {
        inputs: [],
        name: "topTickLowerOrEqBottomTick",
        type: "error",
    },
    {
        inputs: [],
        name: "transferFailed",
        type: "error",
    },
    {
        inputs: [],
        name: "zeroAmountRequired",
        type: "error",
    },
    {
        inputs: [],
        name: "zeroLiquidityActual",
        type: "error",
    },
    {
        inputs: [],
        name: "zeroLiquidityDesired",
        type: "error",
    },
];
class IAlgebraPoolErrors__factory {
    static abi = _abi;
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.IAlgebraPoolErrors__factory = IAlgebraPoolErrors__factory;
