// @cryptoalgebra/integral-core/dist/src/core/test/shared/fixtures.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.poolFixture = exports.TEST_POOL_DAY_BEFORE_START = exports.TEST_POOL_START_TIME = exports.ZERO_ADDRESS = void 0;
const hardhat_1 = require("hardhat");
const ethers_1 = require("ethers");
exports.ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';
async function factoryFixture() {
    const [deployer] = await hardhat_1.ethers.getSigners();
    // precompute
    const poolDeployerAddress = (0, ethers_1.getCreateAddress)({
        from: deployer.address,
        nonce: (await hardhat_1.ethers.provider.getTransactionCount(deployer.address)) + 1,
    });
    const factoryFactory = await hardhat_1.ethers.getContractFactory('AlgebraFactory');
    const factory = (await factoryFactory.deploy(poolDeployerAddress));
    const poolDeployerFactory = await hardhat_1.ethers.getContractFactory('AlgebraPoolDeployer');
    const poolDeployer = (await poolDeployerFactory.deploy(factory));
    const vaultFactory = await hardhat_1.ethers.getContractFactory('AlgebraCommunityVault');
    const vault = (await vaultFactory.deploy(factory, deployer.address));
    const vaultFactoryStubFactory = await hardhat_1.ethers.getContractFactory('AlgebraVaultFactoryStub');
    const vaultFactoryStub = await vaultFactoryStubFactory.deploy(vault);
    await factory.setVaultFactory(vaultFactoryStub);
    return { factory, vault };
}
async function tokensFixture() {
    const tokenFactory = await hardhat_1.ethers.getContractFactory('TestERC20');
    const tokenA = (await tokenFactory.deploy(2n ** 255n));
    const tokenB = (await tokenFactory.deploy(2n ** 255n));
    const tokenC = (await tokenFactory.deploy(2n ** 255n));
    tokenA.address_ = await tokenA.getAddress();
    tokenB.address_ = await tokenB.getAddress();
    tokenC.address_ = await tokenC.getAddress();
    const [token0, token1, token2] = [tokenA, tokenB, tokenC].sort((tokenA, tokenB) => tokenA.address_.toLowerCase() < tokenB.address_.toLowerCase() ? -1 : 1);
    return { token0, token1, token2 };
}
// Monday, October 5, 2020 9:00:00 AM GMT-05:00
exports.TEST_POOL_START_TIME = 1601906400;
exports.TEST_POOL_DAY_BEFORE_START = 1601906400 - 24 * 60 * 60;
const poolFixture = async function () {
    const { factory, vault } = await factoryFixture();
    const { token0, token1, token2 } = await tokensFixture();
    //const { dataStorage } = await dataStorageFixture();
    const MockTimeAlgebraPoolDeployerFactory = await hardhat_1.ethers.getContractFactory('MockTimeAlgebraPoolDeployer');
    const MockTimeAlgebraPoolFactory = await hardhat_1.ethers.getContractFactory('MockTimeAlgebraPool');
    const calleeContractFactory = await hardhat_1.ethers.getContractFactory('TestAlgebraCallee');
    const routerContractFactory = await hardhat_1.ethers.getContractFactory('TestAlgebraRouter');
    const swapTargetCallee = (await calleeContractFactory.deploy());
    const swapTargetRouter = (await routerContractFactory.deploy());
    return {
        token0,
        token1,
        token2,
        factory,
        vault,
        swapTargetCallee,
        swapTargetRouter,
        createPool: async (firstToken = token0, secondToken = token1) => {
            const mockTimePoolDeployer = (await MockTimeAlgebraPoolDeployerFactory.deploy());
            const ADMIN_ROLE = await factory.POOLS_ADMINISTRATOR_ROLE();
            await factory.grantRole(ADMIN_ROLE, mockTimePoolDeployer);
            await mockTimePoolDeployer.deployMock(factory, firstToken, secondToken);
            const firstAddress = await firstToken.getAddress();
            const secondAddress = await secondToken.getAddress();
            const sortedTokens = BigInt(firstAddress) < BigInt(secondAddress) ? [firstAddress, secondAddress] : [secondAddress, firstAddress];
            const poolAddress = await mockTimePoolDeployer.computeAddress(sortedTokens[0], sortedTokens[1]);
            return MockTimeAlgebraPoolFactory.attach(poolAddress);
        },
    };
};
exports.poolFixture = poolFixture;
