// @cryptoalgebra/integral-core/typechain/factories/contracts/interfaces/IAlgebraPoolDeployer__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Interface, type ContractRunner } from "ethers";
import type {
  IAlgebraPoolDeployer,
  IAlgebraPoolDeployerInterface,
} from "../../../contracts/interfaces/IAlgebraPoolDeployer";

const _abi = [
  {
    inputs: [
      {
        internalType: "address",
        name: "plugin",
        type: "address",
      },
      {
        internalType: "address",
        name: "token0",
        type: "address",
      },
      {
        internalType: "address",
        name: "token1",
        type: "address",
      },
      {
        internalType: "address",
        name: "deployer",
        type: "address",
      },
    ],
    name: "deploy",
    outputs: [
      {
        internalType: "address",
        name: "pool",
        type: "address",
      },
    ],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [],
    name: "getDeployParameters",
    outputs: [
      {
        internalType: "address",
        name: "plugin",
        type: "address",
      },
      {
        internalType: "address",
        name: "factory",
        type: "address",
      },
      {
        internalType: "address",
        name: "token0",
        type: "address",
      },
      {
        internalType: "address",
        name: "token1",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
] as const;

export class IAlgebraPoolDeployer__factory {
  static readonly abi = _abi;
  static createInterface(): IAlgebraPoolDeployerInterface {
    return new Interface(_abi) as IAlgebraPoolDeployerInterface;
  }
  static connect(
    address: string,
    runner?: ContractRunner | null
  ): IAlgebraPoolDeployer {
    return new Contract(
      address,
      _abi,
      runner
    ) as unknown as IAlgebraPoolDeployer;
  }
}
