// @cryptoalgebra/integral-core/typechain/contracts/test/FaultyVaultFactoryStub.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  AddressLike,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedListener,
  TypedContractMethod,
} from "../../common";

export interface FaultyVaultFactoryStubInterface extends Interface {
  getFunction(
    nameOrSignature:
      | "createVaultForPool"
      | "defaultAlgebraCommunityVault"
      | "getVaultForPool"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "createVaultForPool",
    values: [AddressLike, AddressLike, AddressLike, AddressLike, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "defaultAlgebraCommunityVault",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "getVaultForPool",
    values: [AddressLike]
  ): string;

  decodeFunctionResult(
    functionFragment: "createVaultForPool",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "defaultAlgebraCommunityVault",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getVaultForPool",
    data: BytesLike
  ): Result;
}

export interface FaultyVaultFactoryStub extends BaseContract {
  connect(runner?: ContractRunner | null): FaultyVaultFactoryStub;
  waitForDeployment(): Promise<this>;

  interface: FaultyVaultFactoryStubInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  createVaultForPool: TypedContractMethod<
    [
      arg0: AddressLike,
      arg1: AddressLike,
      arg2: AddressLike,
      arg3: AddressLike,
      arg4: AddressLike
    ],
    [string],
    "view"
  >;

  defaultAlgebraCommunityVault: TypedContractMethod<[], [string], "view">;

  getVaultForPool: TypedContractMethod<[arg0: AddressLike], [string], "view">;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "createVaultForPool"
  ): TypedContractMethod<
    [
      arg0: AddressLike,
      arg1: AddressLike,
      arg2: AddressLike,
      arg3: AddressLike,
      arg4: AddressLike
    ],
    [string],
    "view"
  >;
  getFunction(
    nameOrSignature: "defaultAlgebraCommunityVault"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "getVaultForPool"
  ): TypedContractMethod<[arg0: AddressLike], [string], "view">;

  filters: {};
}
