// @cryptoalgebra/integral-core/typechain/contracts/test/MockTimeAlgebraPoolDeployer.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  EventFragment,
  AddressLike,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedLogDescription,
  TypedListener,
  TypedContractMethod,
} from "../../common";

export interface MockTimeAlgebraPoolDeployerInterface extends Interface {
  getFunction(
    nameOrSignature:
      | "computeAddress"
      | "deployMock"
      | "getDeployParameters"
      | "mockPoolHash"
  ): FunctionFragment;

  getEvent(nameOrSignatureOrTopic: "PoolDeployed"): EventFragment;

  encodeFunctionData(
    functionFragment: "computeAddress",
    values: [AddressLike, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "deployMock",
    values: [AddressLike, AddressLike, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "getDeployParameters",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "mockPoolHash",
    values?: undefined
  ): string;

  decodeFunctionResult(
    functionFragment: "computeAddress",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "deployMock", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "getDeployParameters",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "mockPoolHash",
    data: BytesLike
  ): Result;
}

export namespace PoolDeployedEvent {
  export type InputTuple = [pool: AddressLike];
  export type OutputTuple = [pool: string];
  export interface OutputObject {
    pool: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export interface MockTimeAlgebraPoolDeployer extends BaseContract {
  connect(runner?: ContractRunner | null): MockTimeAlgebraPoolDeployer;
  waitForDeployment(): Promise<this>;

  interface: MockTimeAlgebraPoolDeployerInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  computeAddress: TypedContractMethod<
    [token0: AddressLike, token1: AddressLike],
    [string],
    "view"
  >;

  deployMock: TypedContractMethod<
    [_factory: AddressLike, token0: AddressLike, token1: AddressLike],
    [string],
    "nonpayable"
  >;

  getDeployParameters: TypedContractMethod<
    [],
    [[string, string, string, string]],
    "view"
  >;

  mockPoolHash: TypedContractMethod<[], [string], "view">;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "computeAddress"
  ): TypedContractMethod<
    [token0: AddressLike, token1: AddressLike],
    [string],
    "view"
  >;
  getFunction(
    nameOrSignature: "deployMock"
  ): TypedContractMethod<
    [_factory: AddressLike, token0: AddressLike, token1: AddressLike],
    [string],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "getDeployParameters"
  ): TypedContractMethod<[], [[string, string, string, string]], "view">;
  getFunction(
    nameOrSignature: "mockPoolHash"
  ): TypedContractMethod<[], [string], "view">;

  getEvent(
    key: "PoolDeployed"
  ): TypedContractEvent<
    PoolDeployedEvent.InputTuple,
    PoolDeployedEvent.OutputTuple,
    PoolDeployedEvent.OutputObject
  >;

  filters: {
    "PoolDeployed(address)": TypedContractEvent<
      PoolDeployedEvent.InputTuple,
      PoolDeployedEvent.OutputTuple,
      PoolDeployedEvent.OutputObject
    >;
    PoolDeployed: TypedContractEvent<
      PoolDeployedEvent.InputTuple,
      PoolDeployedEvent.OutputTuple,
      PoolDeployedEvent.OutputObject
    >;
  };
}
