// @cryptoalgebra/integral-core/typechain/contracts/test/MockDefaultPluginFactory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  EventFragment,
  AddressLike,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedLogDescription,
  TypedListener,
  TypedContractMethod,
} from "../../common";

export interface MockDefaultPluginFactoryInterface extends Interface {
  getFunction(
    nameOrSignature:
      | "afterCreatePoolHook"
      | "beforeCreatePoolHook"
      | "pluginsForPools"
  ): FunctionFragment;

  getEvent(nameOrSignatureOrTopic: "DataOnPoolCreation"): EventFragment;

  encodeFunctionData(
    functionFragment: "afterCreatePoolHook",
    values: [AddressLike, AddressLike, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "beforeCreatePoolHook",
    values: [
      AddressLike,
      AddressLike,
      AddressLike,
      AddressLike,
      AddressLike,
      BytesLike
    ]
  ): string;
  encodeFunctionData(
    functionFragment: "pluginsForPools",
    values: [AddressLike]
  ): string;

  decodeFunctionResult(
    functionFragment: "afterCreatePoolHook",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "beforeCreatePoolHook",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "pluginsForPools",
    data: BytesLike
  ): Result;
}

export namespace DataOnPoolCreationEvent {
  export type InputTuple = [data: BytesLike];
  export type OutputTuple = [data: string];
  export interface OutputObject {
    data: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export interface MockDefaultPluginFactory extends BaseContract {
  connect(runner?: ContractRunner | null): MockDefaultPluginFactory;
  waitForDeployment(): Promise<this>;

  interface: MockDefaultPluginFactoryInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  afterCreatePoolHook: TypedContractMethod<
    [plugin: AddressLike, pool: AddressLike, deployer: AddressLike],
    [void],
    "nonpayable"
  >;

  beforeCreatePoolHook: TypedContractMethod<
    [
      pool: AddressLike,
      arg1: AddressLike,
      arg2: AddressLike,
      arg3: AddressLike,
      arg4: AddressLike,
      data: BytesLike
    ],
    [string],
    "nonpayable"
  >;

  pluginsForPools: TypedContractMethod<[arg0: AddressLike], [string], "view">;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "afterCreatePoolHook"
  ): TypedContractMethod<
    [plugin: AddressLike, pool: AddressLike, deployer: AddressLike],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "beforeCreatePoolHook"
  ): TypedContractMethod<
    [
      pool: AddressLike,
      arg1: AddressLike,
      arg2: AddressLike,
      arg3: AddressLike,
      arg4: AddressLike,
      data: BytesLike
    ],
    [string],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "pluginsForPools"
  ): TypedContractMethod<[arg0: AddressLike], [string], "view">;

  getEvent(
    key: "DataOnPoolCreation"
  ): TypedContractEvent<
    DataOnPoolCreationEvent.InputTuple,
    DataOnPoolCreationEvent.OutputTuple,
    DataOnPoolCreationEvent.OutputObject
  >;

  filters: {
    "DataOnPoolCreation(bytes)": TypedContractEvent<
      DataOnPoolCreationEvent.InputTuple,
      DataOnPoolCreationEvent.OutputTuple,
      DataOnPoolCreationEvent.OutputObject
    >;
    DataOnPoolCreation: TypedContractEvent<
      DataOnPoolCreationEvent.InputTuple,
      DataOnPoolCreationEvent.OutputTuple,
      DataOnPoolCreationEvent.OutputObject
    >;
  };
}
