// @cryptoalgebra/integral-core/typechain/contracts/test/PriceMovementMathTest.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumberish,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedListener,
  TypedContractMethod,
} from "../../common";

export interface PriceMovementMathTestInterface extends Interface {
  getFunction(
    nameOrSignature:
      | "getGasCostOfmovePriceTowardsTarget"
      | "movePriceTowardsTarget"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "getGasCostOfmovePriceTowardsTarget",
    values: [
      BigNumberish,
      BigNumberish,
      BigNumberish,
      BigNumberish,
      BigNumberish
    ]
  ): string;
  encodeFunctionData(
    functionFragment: "movePriceTowardsTarget",
    values: [
      BigNumberish,
      BigNumberish,
      BigNumberish,
      BigNumberish,
      BigNumberish
    ]
  ): string;

  decodeFunctionResult(
    functionFragment: "getGasCostOfmovePriceTowardsTarget",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "movePriceTowardsTarget",
    data: BytesLike
  ): Result;
}

export interface PriceMovementMathTest extends BaseContract {
  connect(runner?: ContractRunner | null): PriceMovementMathTest;
  waitForDeployment(): Promise<this>;

  interface: PriceMovementMathTestInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  getGasCostOfmovePriceTowardsTarget: TypedContractMethod<
    [
      sqrtP: BigNumberish,
      sqrtPTarget: BigNumberish,
      liquidity: BigNumberish,
      amountRemaining: BigNumberish,
      feePips: BigNumberish
    ],
    [bigint],
    "view"
  >;

  movePriceTowardsTarget: TypedContractMethod<
    [
      sqrtP: BigNumberish,
      sqrtPTarget: BigNumberish,
      liquidity: BigNumberish,
      amountRemaining: BigNumberish,
      feePips: BigNumberish
    ],
    [
      [bigint, bigint, bigint, bigint] & {
        sqrtQ: bigint;
        amountIn: bigint;
        amountOut: bigint;
        feeAmount: bigint;
      }
    ],
    "view"
  >;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "getGasCostOfmovePriceTowardsTarget"
  ): TypedContractMethod<
    [
      sqrtP: BigNumberish,
      sqrtPTarget: BigNumberish,
      liquidity: BigNumberish,
      amountRemaining: BigNumberish,
      feePips: BigNumberish
    ],
    [bigint],
    "view"
  >;
  getFunction(
    nameOrSignature: "movePriceTowardsTarget"
  ): TypedContractMethod<
    [
      sqrtP: BigNumberish,
      sqrtPTarget: BigNumberish,
      liquidity: BigNumberish,
      amountRemaining: BigNumberish,
      feePips: BigNumberish
    ],
    [
      [bigint, bigint, bigint, bigint] & {
        sqrtQ: bigint;
        amountIn: bigint;
        amountOut: bigint;
        feeAmount: bigint;
      }
    ],
    "view"
  >;

  filters: {};
}
