// @cryptoalgebra/integral-core/typechain/contracts/test/TestAlgebraRouter.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumberish,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  EventFragment,
  AddressLike,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedLogDescription,
  TypedListener,
  TypedContractMethod,
} from "../../common";

export interface TestAlgebraRouterInterface extends Interface {
  getFunction(
    nameOrSignature:
      | "algebraSwapCallback"
      | "swapForExact0Multi"
      | "swapForExact1Multi"
  ): FunctionFragment;

  getEvent(nameOrSignatureOrTopic: "SwapCallback"): EventFragment;

  encodeFunctionData(
    functionFragment: "algebraSwapCallback",
    values: [BigNumberish, BigNumberish, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "swapForExact0Multi",
    values: [AddressLike, AddressLike, AddressLike, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "swapForExact1Multi",
    values: [AddressLike, AddressLike, AddressLike, BigNumberish]
  ): string;

  decodeFunctionResult(
    functionFragment: "algebraSwapCallback",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "swapForExact0Multi",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "swapForExact1Multi",
    data: BytesLike
  ): Result;
}

export namespace SwapCallbackEvent {
  export type InputTuple = [
    amount0Delta: BigNumberish,
    amount1Delta: BigNumberish
  ];
  export type OutputTuple = [amount0Delta: bigint, amount1Delta: bigint];
  export interface OutputObject {
    amount0Delta: bigint;
    amount1Delta: bigint;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export interface TestAlgebraRouter extends BaseContract {
  connect(runner?: ContractRunner | null): TestAlgebraRouter;
  waitForDeployment(): Promise<this>;

  interface: TestAlgebraRouterInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  algebraSwapCallback: TypedContractMethod<
    [amount0Delta: BigNumberish, amount1Delta: BigNumberish, data: BytesLike],
    [void],
    "nonpayable"
  >;

  swapForExact0Multi: TypedContractMethod<
    [
      recipient: AddressLike,
      poolInput: AddressLike,
      poolOutput: AddressLike,
      amount0Out: BigNumberish
    ],
    [void],
    "nonpayable"
  >;

  swapForExact1Multi: TypedContractMethod<
    [
      recipient: AddressLike,
      poolInput: AddressLike,
      poolOutput: AddressLike,
      amount1Out: BigNumberish
    ],
    [void],
    "nonpayable"
  >;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "algebraSwapCallback"
  ): TypedContractMethod<
    [amount0Delta: BigNumberish, amount1Delta: BigNumberish, data: BytesLike],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "swapForExact0Multi"
  ): TypedContractMethod<
    [
      recipient: AddressLike,
      poolInput: AddressLike,
      poolOutput: AddressLike,
      amount0Out: BigNumberish
    ],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "swapForExact1Multi"
  ): TypedContractMethod<
    [
      recipient: AddressLike,
      poolInput: AddressLike,
      poolOutput: AddressLike,
      amount1Out: BigNumberish
    ],
    [void],
    "nonpayable"
  >;

  getEvent(
    key: "SwapCallback"
  ): TypedContractEvent<
    SwapCallbackEvent.InputTuple,
    SwapCallbackEvent.OutputTuple,
    SwapCallbackEvent.OutputObject
  >;

  filters: {
    "SwapCallback(int256,int256)": TypedContractEvent<
      SwapCallbackEvent.InputTuple,
      SwapCallbackEvent.OutputTuple,
      SwapCallbackEvent.OutputObject
    >;
    SwapCallback: TypedContractEvent<
      SwapCallbackEvent.InputTuple,
      SwapCallbackEvent.OutputTuple,
      SwapCallbackEvent.OutputObject
    >;
  };
}
